THE-app/app/(tabs)/booksScreen.tsx

```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import { ScrollView, StyleSheet } from 'react-native';
+import { Text, View } from '@/components/Themed';
+import BookCard from '@/components/BookCard';
+import { Book } from '@/types/Book';
+
+import { gql, useQuery } from '@apollo/client';
+
+const GET_BOOKS = gql`
+  query {
+    viewer {
+      books {
+        hits {
+          id
+          displayTitle
+          url
+          description
+          subjects {
+            name
+          }
+          levels {
+            name
+          }
+          valid
+        }
+      }
+    }
+  }
+`;
+
+function BooksScreen() {
+  const { loading, error, data } = useQuery(GET_BOOKS);
+
+  if (loading) return <Text>Loading...</Text>;
+  if (error) return <Text>Error :( {error.message}</Text>;
+
+  const books: Book[] = data.viewer.books.hits.filter((book: Book) => book.valid);
```

Comment:
I'm surprised this works, since `data` would initially be undefined.

It's also important in React to `useMemo` for non-primitive types, to avoid unnecessary rerenders.

Reply:
After trying to `useMemo` like that instead :
```
const books: Book[] = useMemo(() => {
    return data.viewer.books.hits.filter((book: Book) => book.valid) || [];
  }, [data]);
```
And getting the following error : 
```
React has detected a change in the order of Hooks called by BooksScreen . 
This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks

   Previous render            Next render
   ------------------------------------------------------
1. useState                   useState
2. useEffect                  useEffect
3. useContext                 useContext
4. useContext                 useContext
5. useRef                     useRef
6. useReducer                 useReducer
7. useContext                 useContext
8. useMemo                    useMemo
9. useCallback                useCallback
10. useSyncExternalStore      useSyncExternalStore
11. undefined                 useMemo
```
I may have an explanation of why it's working, since this part of code wasn't executed while data was undefined, it doesn't cause issues.